### PR TITLE
Feature/download image

### PR DESCRIPTION
**Pull request by cassand**

### CODE REVIEW

File: backend/app/modules/imagefetcher/FetcherActor.scala [hunk 1]
```diff
@@ -85,30 +88,36 @@ class FetcherActor(ws: WSClient, serverUrl: String, monitoring: ActorRef) extend
   }
 
   /**
-    * Download a file from an url through an akka stream an place it a file
-    * (Downloaded file is not labeled yet)
+    * Download a file from an url through an akka stream an place it in the downloaded folder
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/FetcherActor.scala]: s/an place/and place/g

File: backend/app/modules/imagefetcher/FetcherActor.scala [hunk 2]
```diff
@@ -85,30 +88,36 @@ class FetcherActor(ws: WSClient, serverUrl: String, monitoring: ActorRef) extend
...
+    * (Downloaded file is not labeled yet e.g random name)
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/FetcherActor.scala]: s/ e.g random name/; we use a random id/g

File: backend/app/modules/imagefetcher/FetcherActor.scala [hunk 3]
```diff
@@ -85,30 +88,36 @@ class FetcherActor(ws: WSClient, serverUrl: String, monitoring: ActorRef) extend
...
     */
-  def downloadFile(url: String) = {
+  def downloadFile(url: String, processId: String) = {
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/FetcherActor.scala]: undocumented parameters

File: backend/app/modules/imagefetcher/FetcherActor.scala [hunk 4]
```diff
@@ -85,30 +88,36 @@ class FetcherActor(ws: WSClient, serverUrl: String, monitoring: ActorRef) extend
...
 
     val currentSender = sender
 
-    val file = new File("Downloaded/" + Random.nextInt())
+    val downloaded = new File("Downloaded").mkdir()
+    val file = new File("Downloaded/" + Random.nextInt() + ".zip")
     val response = ws.url(url).withMethod("GET").stream()
 
     val downloadedFile = response.flatMap(res => {
+
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/FetcherActor.scala:111]: remove useless blank line

File: backend/app/modules/imagefetcher/FetcherActor.scala [hunk 5]
```diff
@@ -85,30 +88,36 @@ class FetcherActor(ws: WSClient, serverUrl: String, monitoring: ActorRef) extend
...
+      task.map(t => monitoring ! UpdateTask(processId, t.id, Some("Download started"), Some(20)))
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/FetcherActor.scala:112]: `Some(20)` ? Why 20 ? From where ?

File: backend/app/modules/imagefetcher/FetcherActor.scala (hunk 2 above)
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/FetcherActor.scala]: By the way, you may want to use the route parameter of the image to store the picture. Two generation of the same picture will generate the same route as much as I tested.

File: backend/app/modules/scheduler/SchedulerActor.scala [hunk 6]
```diff
@@ -48,18 +50,29 @@ class SchedulerActor(processes: Queries, configuration: play.api.Configuration,
     * * Download image
```
Comment (FunkySayu) [inline backend/app/modules/scheduler/SchedulerActor.scala]: Double stars on the comments (I think you should use full english sentence to explain what does the pipeline); id parameter is not documented

File: backend/app/modules/scheduler/SchedulerActor.scala [hunk 7]
```diff
@@ -48,18 +50,29 @@ class SchedulerActor(processes: Queries, configuration: play.api.Configuration,
...
     */
   def startProcessing(id: String): Unit = {
-      monitoring.ask(GetProcess(id))
+      val details = monitoring
```
Comment (FunkySayu) [inline backend/app/modules/scheduler/SchedulerActor.scala]: `details` is a bit obscure here. What is it? Process details? if so, use `process_details`. Or is it download details ? if so, use `download_details`.

File: backend/app/modules/scheduler/SchedulerActor.scala [hunk 8]
```diff
@@ -48,18 +50,29 @@ class SchedulerActor(processes: Queries, configuration: play.api.Configuration,
...
-          case Some(details) => details
+          case Some(detail) => detail
           case _ =>
             monitoring ! UpdateProcess(id, "failed: no details")
             throw new Exception("no details")
         }
-        .flatMap(details => fetcherActor.ask(FetchRGB("2015-01-01", details.place, Some(10), id)))
+
+      val url = details
+        .flatMap(details => fetcherActor.ask(FetchRGB("2015-01-01", details.place, Some(100), id)))
```
Comment (FunkySayu) [inline backend/app/modules/scheduler/SchedulerActor.scala]: Is scale required here (`Some(100)`)? Shouldn't we use the default scale?

File: backend/app/modules/scheduler/SchedulerActor.scala [hunk 9]
```diff
@@ -48,18 +50,29 @@ class SchedulerActor(processes: Queries, configuration: play.api.Configuration,
...
         .mapTo[FetchResponse]
-        .foreach(response => println("Url from falsk server : " + response.url))
+
+      url.foreach(u => println(u.url))
```
Comment (FunkySayu) [inline backend/app/modules/scheduler/SchedulerActor.scala]: A bit odd to see `url.foreach`. What is this? A list of URLs? Add a `s` after `url` then.

Comment (FunkySayu) [inline backend/app/modules/scheduler/SchedulerActor.scala]: Also mark this as debug (`TODO(xxx): This is for debug purpose`)

File: backend/app/modules/scheduler/SchedulerActor.scala [hunk 10]
```diff
@@ -48,18 +50,29 @@ class SchedulerActor(processes: Queries, configuration: play.api.Configuration,
...
+
+      val file = url
+        .flatMap(response => fetcherActor.ask(DownloadFile(response.url, id)))
+        .mapTo[Downloaded]
+
+      file foreach (downloaded => println(downloaded.file.getAbsolutePath))
```
Comment (FunkySayu) [inline backend/app/modules/scheduler/SchedulerActor.scala]: Same here, mark this as debug.

File: backend/app/modules/imagefetcher/ZipUtil.scala [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    * @param extension A string which is matched against file in the zip to determine which one to extract
+    * @param outputFolder Folder in which the file will be extracted
+    * @return An optional result containing a path to the first file extracted
+    */
+  def extractZip(file: File, extension: String = ".png", outputFolder: String = "./Downloaded"): Option[File] = {
+
+    val zipStream = new ZipInputStream(new FileInputStream(file))
+
+    while(zipStream.available() != 0) {
+
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala:23]: remove blank line

File: backend/app/modules/imagefetcher/ZipUtil.scala [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      val zipEntry = zipStream.getNextEntry
+
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala]: remove blank line

File: backend/app/modules/imagefetcher/ZipUtil.scala [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+      val filename = zipEntry.getName
+
+      if (filename.endsWith(extension)) {
+
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala:28]: remove blank line

File: backend/app/modules/imagefetcher/ZipUtil.scala [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    * @param file File object representing the zip file
+    * @param extension A string which is matched against file in the zip to determine which one to extract
+    * @param outputFolder Folder in which the file will be extracted
+    * @return An optional result containing a path to the first file extracted
+    */
+  def extractZip(file: File, extension: String = ".png", outputFolder: String = "./Downloaded"): Option[File] = {
+
+    val zipStream = new ZipInputStream(new FileInputStream(file))
+
+    while(zipStream.available() != 0) {
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala]: add space after `while`

File: backend/app/modules/imagefetcher/ZipUtil.scala [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+      val zipEntry = zipStream.getNextEntry
+
+      val filename = zipEntry.getName
+
+      if (filename.endsWith(extension)) {
+
+        val outputFile = new File(outputFolder + File.separator + filename)
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala]: move this after the comment, and remove the blank line

File: backend/app/modules/imagefetcher/ZipUtil.scala [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+        // Extract the image from the archive
+        val outputStream: FileOutputStream = new FileOutputStream(outputFile)
+        var size: Int = 0
+        val buffer = new Array[Byte](1024)
+
+        size = zipStream.read(buffer)
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala]: use `val size`, and remove the above `var size: Int = 0` definition.

File: backend/app/modules/imagefetcher/ZipUtil.scala [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+        try {
+
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala]: remove blank line

File: backend/app/modules/imagefetcher/ZipUtil.scala [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+          while (size != 0) {
+            outputStream.write(buffer, 0, size)
+            size = zipStream.read(buffer)
+          }
+        } catch {
+          case e: Throwable => println(e)
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala]: use `err` or `exception` instead of `e`. Much more explicit, even if it is understandable :)

Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala]: Shouldn't this be reported somewhere somehow ?

File: backend/app/modules/imagefetcher/ZipUtil.scala [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        outputStream.close()
+        return Some(outputFile)
+      }
+
+    }
+
+    None
+
+  }
+}
```
Comment (FunkySayu) [inline backend/app/modules/imagefetcher/ZipUtil.scala:45]: Add blank line above

File: backend/app/modules/scheduler/SchedulerActor.scala [hunk 20]
```diff
@@ -48,18 +50,35 @@ class SchedulerActor(processes: Queries, configuration: play.api.Configuration,
...
             monitoring ! UpdateProcess(id, "failed: no details")
             throw new Exception("no details")
         }
-        .flatMap(details => fetcherActor.ask(FetchRGB("2015-01-01", details.place, Some(10), id)))
-        .mapTo[FetchResponse]
-        .foreach(response => println("Url from falsk server : " + response.url))
+
+    // Send processing request to Google Earth Engine using the python server
+    val url = details
+      .flatMap(details => fetcherActor.ask(FetchRGB("2015-01-01", details.place, Some(100), id)))
```
Comment (FunkySayu) [inline backend/app/modules/scheduler/SchedulerActor.scala]: Is scale required here (Some(100))? Shouldn't we use the default scale?

**Review by FunkySayu — CHANGES_REQUESTED** [review]
Some of my comments where outdated on the previous version (I forgot a commit, ups).

File: backend/app/modules/imagefetcher/FetcherActor.scala (hunk 5 above)
Comment (cassand) [reply backend/app/modules/imagefetcher/FetcherActor.scala:112]: This is the task progression, therefore starting download <=> 20%

**Review by FunkySayu — APPROVED** [review]

**Comment by FunkySayu** [comment]
Nice job!

Generally speaking, avoid putting too much blank line in your code. This means less lines on the screen and it sometimes isn't useful / relevant to break the code.
On flow control statement (such as `while`, `if`...), you should avoid putting empty line right after, unless the statement is on several line (and in that specific case, you should refactor your code ;) ). Indentation is pretty meaningful at this point.

This `Some(20)` is still a bit obscure, but I don't see anything else to do so...

**Pull request merged**